### PR TITLE
Fix #8188: For the purposes of overloading, contextual methods aren't methodic

### LINKED ISSUE #8188: Overloaded extension methods with context parameters cannot be fully constructed
## minimized code

```Scala
object Test {
  extension StrDeco on (tree: String) {
    def show(given DummyImplicit): String = ???
    def show(color: Boolean)(given DummyImplicit): String = ???
  }

  val a: String = "foo".show
  val b: String = "foo".show(true)
  val c: Any = "foo".show
  val d: Any = "foo".show(true)
}
```

## Compilation output

```scala
11 |  val c: Any = "foo".show
   |               ^^^^^^^^^^
   |      value show is not a member of String.
   |      An extension method was tried, but could not be fully constructed:
   |
   |          Test.StrDeco.show("foo")
```

## expectation
It should compile

### PR DESCRIPTION
**Pull request by anatoliykmetyuk**

### CODE REVIEW

**Comment by anatoliykmetyuk** [comment]
In the issue, there are two methods for the overloading resolution to try:

```scala
    def show(str: String)(given DummyImplicit): String = ???
    def show(str: String)(color: Boolean)(given DummyImplicit): String = ???
```

After application to `str`, the result types are:

```scala
(given DummyImplicit): String
(color: Boolean)(given DummyImplicit): String 
```

The overloading mechanism prefers non-methodic result types. One can argue that contextual functions are non-methodic because they are not meant to be used as methods.

**Comment by anatoliykmetyuk** [comment]
Incidentally, also, on master:

```scala
scala> class Foo {
     |   def f(x: String)(given DummyImplicit): String = x * 2
     |   def f(x: String)(color: Boolean)(given DummyImplicit): String = x * 3
     | }
// defined class Foo

scala> new Foo().f("foo")
1 |new Foo().f("foo")
  |^^^^^^^^^^^
  |Ambiguous overload. The overloaded alternatives of method f in class Foo with types
  | (x: String)(color: Boolean)(given x$3: DummyImplicit): String
  | (x: String)(given x$2: DummyImplicit): String
  |both match arguments (("foo" : String))
```

On this PR:

```scala
scala> class Foo {
     |   def f(x: String)(given DummyImplicit): String = x * 2
     |   def f(x: String)(color: Boolean)(given DummyImplicit): String = x * 3
     | }
// defined class Foo

scala> new Foo().f("foo")
val res0: String = foofoo
```

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 1]
```diff
@@ -1783,7 +1783,10 @@ trait Applications extends Compatibility {
     }
 
     def resultIsMethod(tp: Type): Boolean = tp.widen.stripPoly match
-      case tp: MethodType => tp.resultType.isInstanceOf[MethodType]
+      case tp: MethodType =>
+        tp.resultType match
+          case mt: MethodType => !mt.isContextualMethod
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala]: That does not handle the case correctly where a context method type has a regular method type as a result. You can do this instead:
```scala
case tp: MethodType => tp.resultType.stripImplicit.isInstanceOf[MethodType]
```

**Comment by anatoliykmetyuk** [comment]
The following test fails to compile: https://github.com/lampepfl/dotty/blob/228b61ea174e1768089d957538cbeb91e35dd76a/tests/pos/i2378.scala

The error message says:

```scala
-- [E127] Syntax Error: /Users/kmetiuk/Projects/scala3/pg/i8188/iss_f4.scala:7:7
7 |  case Apply(fun) => 3
  |       ^^^^^
  |Apply cannot be used as an extractor in a pattern because it lacks an unapply or unapplySeq method

longer explanation available when compiling with `-explain`
1 error found
```

Also, currently, on `master`, the following compiles:

```scala
object Apply {
  def unapply(tree: String): Option[String] = { println(s"No implicit"); Some(tree) }
  def unapply(tree: Int)(implicit c: DummyImplicit): Option[String] = { println("Implicit"); Some(tree.toString) }
}

def foo(tree: Any): Int = tree match {
  case Apply(fun) => 3
}

@main def Test = foo(2)
```

On runtime it fails with:

```scala
Exception in thread "main" scala.MatchError: 2 (of class java.lang.Integer)
	at iss_f4$package$.foo(iss_f4.scala:7)
	at iss_f4$package$.Test(iss_f4.scala:10)
	at Test.main(iss_f4.scala:10)
```

Because the first `unapply` is always preferred. Scala 2 rejects the code above with the following message:

```scala
scala> def foo(tree: Any): Int = tree match {
     |   case Apply(fun) => 3
     | }
         case Apply(fun) => 3
              ^
On line 2: error: cannot resolve overloaded unapply
```

The failing test in question was previously also failing on Dotty, up until https://github.com/lampepfl/dotty/commit/d8bcc16e1f20933fb8d26ff1f38831fd1b7a0972 when it was moved from `neg` to `pos`.

I think the failing behavior for the test is correct – the pattern match in question is:

```scala
  def foo(tree: Tree): Int = (tree: Any) match {
    case tb.Apply(fun, args) => 3
  }
```

The scrutinee is of type `Any`, hence both of the `unapply` methods can be used as extractor on it. However, due to the rule introduced by the above commit, we now prefer methods that are not curried during overload resoltion. This rule gives priority to the methods without implicit parameters. It leads to runtime crashes as in case of `@main def Test = foo(2)` in the minimized example above.

I think the behavior here should be as in Scala 2: report an overload resolution error if there are two `unapply`s that can be used, even if one of them accepts implicit parameters. This PR achieves this error report by treating contextual methods as non-methodic. However, the error message needs work, maybe in this PR or maybe in another: it shouldn't say that there are no `unapply` methods defined in the object but it should instead report an overload resolution error.

WDYT @odersky?

**Comment by odersky** [comment]
Is this still an issue when #8406 is in?

**Comment by anatoliykmetyuk** [comment]
Yes, this seems to be a different issue.

**Pull request merged**